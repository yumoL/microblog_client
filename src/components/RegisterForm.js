/**
 * @description register form
 */
import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { Grid, Header, Form, Segment, Button, Message } from 'semantic-ui-react'
import { Link, withRouter, Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import { checkUserNameExisted, register } from '../reducers/userReducer'
import { useDebounce } from '../hooks/useDebounce'
import { useField } from '../hooks/useField'
import { setNotification } from '../reducers/notificationReducer'
import ValidateInput from './ValidateInput'
import RedirectToMain from './RedirectToMain'


const RegisterForm = props => {
  const [registerSuccess, setRegisterSuccess] = useState(false)
  const userName = useField('text')
  const password = useField('password')
  const passwordConfirm = useField('password')
  const debouncedUsername = useDebounce(userName.value, 500)

  useEffect(() => {
    if (debouncedUsername) {
      const existedUserName = async () => {
        await props.checkUserNameExisted(debouncedUsername)
      }
      existedUserName()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedUsername])

  const submitButtonDisabled = props.user.userNameExisted || userName.value.length < 3 || userName.value.length > 255 || password.value.length < 3 || password.value.length > 255 || password.value !== passwordConfirm.value

  const handleSubmit = async e => {
    e.preventDefault()
    const newUser = {
      userName: userName.value,
      password: passwordConfirm.value
    }
    try {
      await props.register(newUser)
      setRegisterSuccess(true)
      //props.history.push('/login')
      props.setNotification({ message: 'Register succeeded, please log in', error: false })
    } catch (e) {
      props.setNotification({ message: 'Register failed', error: true })
    }
  }

  if (registerSuccess) {
    return (
      <div className='redirect'>
        <Redirect to='/login' />
      </div>
    )
  }

  if (props.user.token) {
    return (
      <RedirectToMain />
    )
  }

  return (
    <div>
      <Grid textAlign='center' style={{ height: '100vh' }} verticalAlign='middle'>
        <Grid.Column style={{ maxWidth: 450 }}>
          <Header as='h2' color='teal' textAlign='center'>
            Sign Up
          </Header>
          <Form size='large' onSubmit={handleSubmit}>
            <Segment stacked>
              <Form.Input id='userName' fluid icon='user' iconPosition='left' placeholder='Username (between 3 and 255 characters)'
                {...userName} required
              />
              <ValidateInput fieldName={'Username'} input={userName.value} minLength={3} maxLength={255}
                extraRequirement={!props.user.userNameExisted} extraErrorMessage={'User name is already existed'} />
              <Form.Input
                id='pwd'
                fluid
                icon='lock'
                iconPosition='left'
                placeholder='Password (between 3 and 255 characters)'
                {...password}
                required
              />
              <ValidateInput fieldName={'Password'} input={password.value} minLength={3} maxLength={255} />
              <Form.Input
                id='pwdConfirm'
                fluid
                placeholder='Confirm Your Password'
                {...passwordConfirm}
                required
              />
              <ValidateInput fieldName={'Password confirmation'} input={passwordConfirm.value}
                extraRequirement={password.value === passwordConfirm.value} extraErrorMessage={'Confirmation does not match the password'} />
              <Button color='teal' fluid size='large' disabled={submitButtonDisabled}>
                Sign Up
              </Button>
            </Segment>
          </Form>
          <Message>Already have an account? <Link to='/login'>Back to login</Link></Message>
        </Grid.Column>
      </Grid>

    </div>
  )
}

const mapStateToProps = state => {
  return {
    user: state.user,
    file: state.file
  }
}

const mapDispatchToProps = {
  checkUserNameExisted,
  register,
  setNotification
}

RegisterForm.propTypes = {
  history: PropTypes.object.isRequired,
  checkUserNameExisted: PropTypes.func.isRequired,
  register: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  setNotification: PropTypes.func.isRequired
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(RegisterForm)
)
