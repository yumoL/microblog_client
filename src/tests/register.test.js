/* eslint-disable no-undef */
import React from 'react'
import RegisterForm from '../components/RegisterForm'
import { renderWithReduxAndRouter } from './testHelper'
import { fireEvent, wait } from '@testing-library/react'
jest.mock('../services/user')

describe('Register', () => {
  let Wrapper
  let userNameInput
  let pwdInput
  let pwdConfirmInput

  beforeEach(() => {
    Wrapper = renderWithReduxAndRouter(<RegisterForm />)
    userNameInput = Wrapper.container.querySelector('.userName').querySelector('input')
    pwdInput = Wrapper.container.querySelector('.pwd').querySelector('input')
    pwdConfirmInput = Wrapper.container.querySelector('.pwdConfirm').querySelector('input')
  })

  test('should render register form correctly', () => {
    expect(Wrapper).toMatchSnapshot()
  })

  describe('Show warning message when input is invalid', () => {
    test('Username is already existed', async () => {
      fireEvent.change(userNameInput, { target: { value: 'user1' } })
      await wait(() => expect(Wrapper.getByText('User name is already existed')).toBeInTheDocument())
    })

    test('Password confirmation does not match the password input', () => {
      fireEvent.change(pwdInput, { target: { value: 'pwd' } })
      fireEvent.change(pwdConfirmInput, { target: { value: 'unmatchedPwd' } })
      expect(Wrapper.getByText('Confirmation does not match the password')).toBeInTheDocument()
    })
  })

  test('Redirect to login page after successful registration', async () => {
    const form = Wrapper.container.querySelector('form')
    fireEvent.change(userNameInput, { target: { value: 'user1' } })
    fireEvent.change(pwdInput, { target: { value: 'pwd' } })
    fireEvent.change(pwdConfirmInput, { target: { value: 'pwd' } })

    fireEvent.submit(form)

    await wait(() => expect(Wrapper.container.querySelector('.redirect')).toBeInTheDocument())
    expect(Wrapper.container.querySelector('.pwdConfirm')).not.toBeTruthy()
  })
})