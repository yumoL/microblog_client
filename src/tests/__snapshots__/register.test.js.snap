// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render register form correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="ui center aligned middle aligned grid"
          style="height: 100vh;"
        >
          <div
            class="column"
            style="max-width: 450px;"
          >
            <h2
              class="ui teal center aligned header"
            >
              Sign Up
            </h2>
            <form
              class="ui large form"
            >
              <div
                class="ui stacked segment"
              >
                <div
                  class="required field userName"
                >
                  <div
                    class="ui fluid left icon input"
                  >
                    <input
                      placeholder="Username (between 3 and 255 characters)"
                      required=""
                      type="text"
                      value=""
                    />
                    <i
                      aria-hidden="true"
                      class="user icon"
                    />
                  </div>
                </div>
                <div
                  class="required field pwd"
                >
                  <div
                    class="ui fluid left icon input"
                  >
                    <input
                      placeholder="Password (between 3 and 255 characters)"
                      required=""
                      type="password"
                      value=""
                    />
                    <i
                      aria-hidden="true"
                      class="lock icon"
                    />
                  </div>
                </div>
                <div
                  class="required field pwdConfirm"
                >
                  <div
                    class="ui fluid input"
                  >
                    <input
                      placeholder="Confirm Your Password"
                      required=""
                      type="password"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="ui label"
                >
                  Upload Your Head Portrait
                </div>
                <div
                  class="field"
                >
                  <div
                    class="ui input"
                  >
                    <input
                      type="file"
                    />
                  </div>
                </div>
                <button
                  class="ui teal large fluid disabled button"
                  disabled=""
                  tabindex="-1"
                >
                  Sign Up
                </button>
              </div>
            </form>
            <div
              class="ui message"
            >
              Already have an account? 
              <a
                href="/login"
              >
                Back to login
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="ui center aligned middle aligned grid"
        style="height: 100vh;"
      >
        <div
          class="column"
          style="max-width: 450px;"
        >
          <h2
            class="ui teal center aligned header"
          >
            Sign Up
          </h2>
          <form
            class="ui large form"
          >
            <div
              class="ui stacked segment"
            >
              <div
                class="required field userName"
              >
                <div
                  class="ui fluid left icon input"
                >
                  <input
                    placeholder="Username (between 3 and 255 characters)"
                    required=""
                    type="text"
                    value=""
                  />
                  <i
                    aria-hidden="true"
                    class="user icon"
                  />
                </div>
              </div>
              <div
                class="required field pwd"
              >
                <div
                  class="ui fluid left icon input"
                >
                  <input
                    placeholder="Password (between 3 and 255 characters)"
                    required=""
                    type="password"
                    value=""
                  />
                  <i
                    aria-hidden="true"
                    class="lock icon"
                  />
                </div>
              </div>
              <div
                class="required field pwdConfirm"
              >
                <div
                  class="ui fluid input"
                >
                  <input
                    placeholder="Confirm Your Password"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
              </div>
              <div
                class="ui label"
              >
                Upload Your Head Portrait
              </div>
              <div
                class="field"
              >
                <div
                  class="ui input"
                >
                  <input
                    type="file"
                  />
                </div>
              </div>
              <button
                class="ui teal large fluid disabled button"
                disabled=""
                tabindex="-1"
              >
                Sign Up
              </button>
            </div>
          </form>
          <div
            class="ui message"
          >
            Already have an account? 
            <a
              href="/login"
            >
              Back to login
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
